banyan/src/forest/read.rs

```diff
@@ -14,21 +14,106 @@ use futures::{prelude::*, stream::BoxStream};
 use libipld::cbor::DagCbor;
 use smallvec::{smallvec, SmallVec};
 
-use std::{iter, marker::PhantomData, ops::Range, time::Instant};
+use std::{iter, marker::PhantomData, ops::Range, sync::Arc, time::Instant};
+
+pub(crate) trait TreeVisitor<T: TreeTypes, R> {
+    type Item;
+    /// We are going to skip a branch or leaf. Compute placeholder item.
+    fn skip(&self, range: Range<u64>, index: &NodeInfo<T, R>) -> Self::Item;
+    /// We made it all the way to a leaf. Compute a value from it.
+    ///
+    /// Here we have the choice of loading the leaf or not. Since the loading
+    /// can fail, this method returns a Result.
+    fn leaf(
+        &self,
+        range: Range<u64>,
+        index: Arc<LeafIndex<T>>,
+        leaf: LeafLoader<T, R>,
+        mask: &[bool],
+    ) -> Result<Self::Item>;
+}
+
+/// A tree visitor that produces chunks, consisting of value triples and some
+/// arbitary extra data.
+pub(crate) struct ChunkVisitor<F, X>
+where
+    F: 'static,
+{
+    mk_extra: &'static F,
+    _p: PhantomData<X>,
+}
+
+impl<F, X> ChunkVisitor<F, X> {
+    pub fn new(mk_extra: &'static F) -> Self {
```

Comment:
this `'static` smells weird — can this actually be called with anything but a `fn`?

Reply:
Well, you can call new with anything, but the result won't be a good visitor since it then won't implement TreeVisitor. But that is how it is often done in rust: don't have constraints on the struct but only on the mipls, otherwise you have an insane explosion of constraints all over the place.

Reply:
That’s not what I meant: how do you construct a static reference to a closure? The only way I know is to not use a closure at all, at which point the correct signature would start with `fn`.

Reply:
I still don't know what exactly you suggest. Maybe just make a code suggestion? In any case this is for now an internal API.

Reply:
I’m not suggesting a change, I’m trying to understand how this function is supposed to be called. I guess I’ll findReferences someday and see it.